Convert worker-manager TCP server to an HTTP server for DX

lukemelia opened the pull request.

lukemelia committed:
Convert worker-manager TCP server to an HTTP server that exposes queue depth

lukemelia commented:
@habdelra thoughts about whether and where to test this?

jurgenwerk commented on packages/realm-server/worker-manager.ts:
```diff
@@ -75,63 +81,93 @@ let isExiting = false;
+      ...result,
+      highPriorityWorkers: highPriorityCount,
+      allPriorityWorkers: allPriorityCount,
+      queueDepth: parseInt(queue_depth, 10),
```
Just my personal opinion that `pendingJobs` would be a bit clearer name for this.

jurgenwerk reviewed (approved):
I tested this locally in the browser and I was able to see the stats.

habdelra commented:
So i think the way I would approach testing for this would be to create a new realm server test module for this. in that module I'd set a random DB name using `prepareTestDB()` from the realm-server's test helpers. I would then spawn a child process to start the worker manager in the `beforeEach()`, which in turn will spawn workers and start the HTTP server.. You can hold a reference to the worker manager's process, and in the `afterEach()` kill the worker manager process which should kill the workers since they are subprocesses of the worker manager. since each test uses a different DB, hopefully that will help to accommodate the async used to spin down the worker in case the torn down workers haven't completed torn down by the time the next test starts. Within the body of each test then, you can use the pg queue's publisher to publish jobs to the queue. 

controlling queue depth might be a little tricky for the test. perhaps that means actually you publish jobs to the queue first before you spin up the worker manager--that way you can be sure of how many jobs are in the queue. and after the worker manager spins up you should see the queue depth start to decrease.

habdelra commented:
Also keep in mind that in production right now there are _2_ worker managers. because there are 2 ECS tasks in the worker ECS service. How do you intend to expose the worker manager HTTP for each ECS worker task? For state originating from the DB it probably doesn't matter, since it will be the same--but for worker manager specific state, like number of running workers (and perhaps the PID of the workers) this may not necessarily be the same

habdelra commented on packages/realm-server/worker-manager.ts:
```diff
@@ -75,63 +81,93 @@ let isExiting = false;
+    ])) as {
+      queue_depth: string;
+    }[];
+    result = {
```
another metric that might be nice to report on would be uncompleted jobs in jobs_reservations that have been running longer than some threshold and are not expired

habdelra commented on packages/realm-server/worker-manager.ts:
```diff
@@ -75,63 +81,93 @@ let isExiting = false;
-    shutdown();
-  });
+webServer.on('error', (err: any) => {
+  console.error(`worker manager HTTP server error: ${err.message}`);
```
let's use our logger instead

habdelra commented on packages/realm-server/worker-manager.ts:
```diff
@@ -75,63 +81,93 @@ let isExiting = false;
+      log.error(`Error while closing the server for worker manager HTTP:`, err);
+      process.exit(1);
     }
+    dbAdapter.close(); // warning this is async
```
ignoring this async might make testing trickier.

habdelra commented on packages/realm-server/worker-manager.ts:
```diff
@@ -75,63 +81,93 @@ let isExiting = false;
     }
+    dbAdapter.close(); // warning this is async
+    log.info(`worker manager HTTP on port ${port} has stopped.`);
+    onShutdown?.();
```
consider moving this into the `dbAdapter.close().finally()`...

habdelra commented on packages/realm-server/worker-manager.ts:
```diff
@@ -75,63 +81,93 @@ let isExiting = false;
+    dbAdapter.close(); // warning this is async
+    log.info(`worker manager HTTP on port ${port} has stopped.`);
+    onShutdown?.();
+    process.exit(0);
```
consider moving this also into the `dbAdapter.close().finally()`

lukemelia commented:
After discussion with @habdelra, I am going to update this branch to retain the TCP -> HTTP conversion but drop the queue depth. In a separate PR, I will add an endpoint to realm server to report on queue stats starting with queue depth.

lukemelia committed:
Update to retain the TCP -> HTTP conversion but drop the queue depth.

- In a separate PR, we will add an endpoint to realm server to report on queue stats starting with queue depth.
